hal/critical_section_api.h

```diff
@@ -0,0 +1,53 @@
+/** \addtogroup hal */
+/** @{*/
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MBED_CRITICAL_SECTION_API_H
+#define MBED_CRITICAL_SECTION_API_H
+
+#include <stdbool.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** Mark the start of a critical section
+  *
+  * This function is called directly by core_util_critical_section_enter
+  *
+  * @param  in_nested_critical_section
+  *                     Used to determine if this is the first call to enter a
+  *                     critical section or a nested call. The first call should
+  *                     store the state prior to entering the critical section.
+  */
+void hal_critical_section_enter(const bool in_nested_critical_section);
```

Comment:
Do we need the parameter? Mbed core is keeping count anyway we may as well call it once? What do you think? I would prefer to have complicated drivers and simple HAL implementation @0xc0170 @c1728p9

Reply:
Yeah, that makes more sense. I did it that way because the original implementation let you call enter multiple times and it would re-disable interrupts. But, that doesn't make much sense as it will assert if they've been re-enabled during the critical section.